CesiumGltf/include/CesiumGltf/SharedAssetDepot.h

```diff
@@ -0,0 +1,260 @@
+                    std::shared_ptr<CesiumAsync::IAssetRequest>&& pRequest)
+                    -> CesiumUtility::IntrusivePointer<AssetType> {
+                  const CesiumAsync::IAssetResponse* pResponse =
+                      pRequest->response();
+                  if (!pResponse) {
+                    return nullptr;
+                  }
+
+                  return factory.createFrom(pResponse->data());
+                })
+            // Do this in main thread since we're messing with the collections.
+            .thenInMainThread(
+                [uri,
+                 this,
+                 pAssets = &this->assets,
+                 pPendingAssetsMutex = &this->assetsMutex,
+                 pPendingAssets = &this->pendingAssets](
+                    CesiumUtility::IntrusivePointer<AssetType>&& pResult)
+                    -> CesiumUtility::IntrusivePointer<AssetType> {
+                  std::lock_guard lock(*pPendingAssetsMutex);
+
+                  // Get rid of our future.
+                  pPendingAssets->erase(uri);
+
+                  if (pResult) {
+                    pResult->_pDepot = this;
+                    pResult->_uniqueAssetId = uri;
+
+                    auto [it, ok] = pAssets->emplace(uri, pResult);
+                    if (!ok) {
+                      return nullptr;
+                    }
+
+                    return it->second;
+                  }
+
+                  return nullptr;
+                });
+
+    auto [it, ok] = this->pendingAssets.emplace(uri, std::move(future).share());
+    if (!ok) {
+      return asyncSystem
+          .createResolvedFuture<CesiumUtility::IntrusivePointer<AssetType>>(
+              nullptr)
+          .share();
+    }
+
+    return it->second;
+  }
+
+  /**
+   * @brief Returns the total number of distinct assets contained in this depot.
+   */
+  size_t getDistinctCount() const { return this->assets.size(); }
+
+  /**
+   * @brief Returns the number of active references to assets in this depot.
+   */
+  size_t getUsageCount() const {
+    size_t count = 0;
+    for (const auto& [key, item] : assets) {
+      count += item->_referenceCount;
+    }
+    return count;
+  }
+
+  /**
+   * @brief Obtains statistics on the number of assets currently listed as
+   * candidates for deletion, along with the average number of seconds that
+   * they've been pending deletion.
+   * @param outAverageAge The average time in seconds that the current deletion
+   * candidates have spent pending deletion.
+   * @param outCount The number of current deletion candidates.
+   */
+  void getDeletionStats(float& outAverageAge, size_t& outCount) const {
+    size_t count = 0;
+    float total = 0;
+
+    std::lock_guard lock(this->deletionCandidatesMutex);
+    for (auto& [asset, age] : this->deletionCandidates) {
+      total += age;
+      count++;
+    }
+
+    outAverageAge = count > 0 ? total / count : 0;
+    outCount = count;
+  }
+
+private:
+  // Disable copy
```

Comment:
Can probably just remove this comment, it's not adding anything that isn't obvious.